Tutorial Create a Doctor-centric App with iCure (#6)
icure/medtech-docs: sdks/tutorial/index.md

```

If your user already has its RSA Keypair, you can call `initUserCrypto` providing it, which will only 
store it in the current API. 
Here is an example on how to do it : 
<!-- file://code-samples/tutorial/doctor-centric-app/index.mts snippet:init user crypto with existing key-->
```typescript
const iCureUserPubKey = process.env.ICURE_USER_PUB_KEY!
const iCureUserPrivKey = process.env.ICURE_USER_PRIV_KEY!
await api.initUserCrypto(false, { publicKey: iCureUserPubKey, privateKey: iCureUserPrivKey })
```
For more information about this method, go to the [References: initUserCrypto](../references/classes/MedTechApi.md#initusercrypto)

After instantiating your API, you can now get the information of your logged user. 
<!-- file://code-samples/tutorial/doctor-centric-app/index.mts snippet:get current user-->
```typescript
const loggedUser = await api.userApi.getLoggedUser()
expect(loggedUser.login).to.be.equal(iCureUserLogin)
```

Now that your API is properly initialized and that your user can create encrypted data, let's 
check how you can create a patient with iCure. 

# Create your first patient
In most doctor-centric applications, you will have to create patients.

> To do this, re-use the api you just created and call the `patientApi.createOrModifyPatient` service. 

Author: Yes, you need to

<!-- file://code-samples/tutorial/doctor-centric-app/index.mts snippet:create your first patient-->
```typescript
const createdPatient = await api.patientApi.createOrModifyPatient(new Patient({
    firstName: 'John',
    lastName: 'Snow',
    gender: 'male',
    note: 'Winter is coming'
}))
console.log(`Your new patient id : ${createdPatient.id}`);
```

When creating your patient, iCure will assign an id to it, format as an uuid.

That's exactly what we'll need to retrieve the information of your patient. 

# Get your newly created Patient information

> To retrieve your patient information, you will need to call the `patientApi.getPatient` service, providing
> the id iCure assigned to your patient during its creation. 

Reviewer: ```suggestion
You can retrieve patient information using `patientApi.getPatient` with the id of the patient you want to retrieve.
```

<!-- file://code-samples/tutorial/doctor-centric-app/index.mts snippet:get your patient information-->
```typescript
const johnSnow = await api.patientApi.getPatient(createdPatient.id)
expect(createdPatient.id).to.be.equal(johnSnow.id)
```

If you would like to know more about the information contained in Patient, go check the [References](../references/classes/Patient.md)

Let's now create some medical data for our patient. 

# Create your first data sample

> A data sample can be any piece of information related to the health of a patient. (Measures, symptoms, ...)

Reviewer: I got here will finish tomorrow

as data samples in iCure. 

To create a new data sample, use the `dataSampleApi.createOrModifyDataSample` or `dataSampleApi.createOrModifyDataSamples`
if you have multiple of them. 
You can also link those medical data to the patient you just created through the `patientId`argument. 

<!-- file://code-samples/tutorial/doctor-centric-app/index.mts snippet:create your patient first medical data-->
```typescript
const createdData = await api.dataSampleApi.createOrModifyDataSamplesFor(johnSnow.id, [new DataSample({
    labels: new Set([new CodingReference({ type: 'LOINC', code: '29463-7', version: '2' })]),
    content: { en: { numberValue: 92.5 } },
    valueDate: 20220203111034,
    comment: 'Weight'
}),
    new DataSample({
        labels: new Set([new CodingReference({ type: 'LOINC', code: '8302-2', version: '2' })]),
        content: { 'en': { numberValue: 187 } },
        valueDate: 20220203111034,
        comment: 'Height'
    })
])
```

When creating your data samples, iCure will assign an id to each of them as for patients.

But what happens if you don't know these ids ? How can you find back your data ?

# Search for data samples satisfying some criteria

> Sometimes in your application, you will have to find multiple data, satisfying some specific criteria. 
> Some examples : 
> - You want to find all medical data of your patient with id ...; 
> - You want to find all patients containing 'Smith' in their name; 
> - You want to find all your male patients born between 01-01-1970 and 31-12-1970;
> - ...

Reviewer: ```suggestion
Sometimes you will want to search for data which satisfies some specific criteria. 
For example you may want to find: 
- all medical data of your patient with id `xyz`.
- all patients containing 'Smith' in their name.
- all your male patients born between 01-01-1970 and 31-12-1970.
- ...
```

> For this, we'll use the `dataSampleApi.filterDataSamples` service, creating our own search criteria.
> Let's say we would like to find back all medical data of our previously created Patient, with the corresponding 
> LOINC Body Weight label. 

Reviewer: The examples above mention searching for patients, but here we talk only about data sample filters, and it may be confusing: the user may think they have to use filterDataSamples also to search for patients with some criteria.
It may be better to give some general information and then the example specific for data samples.
```suggestion
You can do this using the `filter` services of the apis, such as `dataSampleApi.filterDataSamples` or `patientApi.filterPatients`.
Let's say we would like to find back all medical data with the LOINC label corresponding to the body weight for the patient we have created. 
```

<!-- file://code-samples/tutorial/doctor-centric-app/index.mts snippet:Find your patient medical data following some criteria-->
```typescript
const johnData = await api.dataSampleApi.filterDataSample(await new DataSampleFilter()
    .forDataOwner(api.dataOwnerApi.getDataOwnerIdOf(loggedUser))
    .forPatients(api.cryptoApi, [johnSnow])
    .byTagCodeFilter('LOINC', '29463-7')
    .build()
)

expect(johnData.rows.length).to.be.equal(1)
expect(johnData.pageSize).to.be.equal(1)
expect(johnData.rows[0].content['en'].numberValue).to.be.equal(92.5)
expect(johnData.rows[0].comment).to.be.equal('Weight')
```

> Filters offer you a lot of possibilities. Go have a look at 
> the [How To search data in iCure using complex filters](../how-to/how-to-filter-data-with-advanced-search-criteria.md) 
> to know more about them. 

Reviewer: ```suggestion
Filters offer you a lot of possibilities. Go have a look at 
the [How To search data in iCure using complex filters](../how-to/how-to-filter-data-with-advanced-search-criteria.md) 
guide to know more about them. 
```

# Get your newly created Data Sample information
Finally, you can also get the information of a specific data sample, if you know its id. 

<!-- file://code-samples/tutorial/doctor-centric-app/index.mts snippet:get specific medical data information-->
```typescript
const johnWeight = await api.dataSampleApi.getDataSample(johnData.rows[0].id)
expect(johnData.rows[0].id).to.be.equal(johnWeight.id)
expect(johnWeight.content['en'].numberValue).to.be.equal(92.5)
expect(johnWeight.comment).to.be.equal('Weight')
```

To know more about the Data Sample information, go check the [References](../references/classes/DataSample.md)

# What's next ? 
Congratulations, you are now able to use the basic functions of the iCure MedTech SDK. 

> However, the SDK offers a lot of other services : head to the [How-To](../how-to/index.md) section to discover all of them. 

Reviewer: ```suggestion
However, the SDK offers a lot of other services: head to the [How-To](../how-to/index.md) section to discover all of them. 
```
